### PR TITLE
add tear down phase and update example

### PR DESCRIPTION
**Pull request by wallacbe**

### CODE REVIEW

**Comment by wallacbe** [comment]
PTAL

File: examples/all_the_things.py [hunk 1]
```diff
@@ -109,5 +111,5 @@ def attachments(test):
   #  with open('my_private_key.json', 'r') as json_file:
   #    test.AddOutputCallbacks(mfg_inspector.UploadToMfgInspector.from_json(
   #        json.load(json_file)))
-
+  test.Configure(teardown_phase=tear_down)
```
Comment (jettisonjoe) [inline examples/all_the_things.py]: Hmm, is teardown/tear_down one word, or two?

**Comment by jettisonjoe** [comment]
My thought is that philosophically, the teardown function is _like_ a phase, but isn't _actually_ a proper phase like other phases. My suggestion is to not called it a phase, but a more generic 'function', and specify in the documentation that we pass it the same args we pass phases for the sake of familiarity.

I'm also thinking this should be done in the exit handler, but could be convinced otherwise.

File: examples/all_the_things.py (hunk 1 above)
Comment (wallacbe) [inline examples/all_the_things.py]: noun=teardown, verb= tear down
I changed all instances to teardown

**Comment by wallacbe** [comment]
Changed all references to "teardown".  Removed mention of phase.  Added mention of function vs phase in Configure().

**Comment by wallacbe** [comment]
Which exit handler do you recommend?  In test.Execute or in TestExecutor?

File: openhtf/exe/__init__.py [hunk 2]
```diff
@@ -209,3 +215,7 @@ def _ExecuteTestPhases(self, executor):
         break
     else:
       self._test_state.SetStateFinished()
+    # Run teardown function.
+    if self._teardown_function:
+      self._test_state.pending_phases.insert(0, self._teardown_function)
```
Comment (jettisonjoe) [inline openhtf/exe/__init__.py]: I would argue there's no need to add this to pending phases if we don't think of it as a phase.

**Comment by jettisonjoe** [comment]
I was thinking in `optionally_stop`. Need @grybmadsci to review as well. Let's see what he thinks.

**Comment by wallacbe** [comment]
I tried moving it to optionally_stopped.  I added a teardown to hello_world.py.  It executes fine normally.  However I also tried adding a delay to hello_world phase and then did CTRL+C.  I see the teardown occur but we get some additional errors:

Provide a DUT ID in order to start the test.
test
Hello World!
^C2016-05-24 15:03:29,164 - ERROR - Received SIGINT.
2016-05-24 15:03:29,165 - ERROR - Stopping Test due to SIGINT: <TestState: test@wallaceb-glaptop Running Phase: hello_world>
teardown
2016-05-24 15:03:29,703 - ERROR - Failed to execute test.
Traceback (most recent call last):
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/__init__.py", line 160, in _ThreadProc
    self._ExecuteTestPhases(executor)
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/__init__.py", line 218, in _ExecuteTestPhases
    for phase_outcome in executor.ExecutePhases():
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_executor.py", line 170, in ExecutePhases
    result = self._ExecuteOnePhase(self.test_state.pending_phases[0])
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_executor.py", line 194, in _ExecuteOnePhase
    outcome_wrapper.SetOutcome(phase_thread.JoinOrDie())
  File "build/bdist.linux-x86_64/egg/contextlib2.py", line 97, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_data.py", line 143, in RecordPhaseTiming
    values = dict(test_state.phase_data.measurements)
TypeError: 'NoneType' object is not iterable
ERROR:root:Thread raised an exception: TestExecutorThread
Traceback (most recent call last):
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/util/threads.py", line 59, in run
    self._ThreadProc()
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/__init__.py", line 160, in _ThreadProc
    self._ExecuteTestPhases(executor)
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/__init__.py", line 218, in _ExecuteTestPhases
    for phase_outcome in executor.ExecutePhases():
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_executor.py", line 170, in ExecutePhases
    result = self._ExecuteOnePhase(self.test_state.pending_phases[0])
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_executor.py", line 194, in _ExecuteOnePhase
    outcome_wrapper.SetOutcome(phase_thread.JoinOrDie())
  File "build/bdist.linux-x86_64/egg/contextlib2.py", line 97, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_data.py", line 143, in RecordPhaseTiming
    values = dict(test_state.phase_data.measurements)
TypeError: 'NoneType' object is not iterable
Exception in thread TestExecutorThread:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/util/threads.py", line 59, in run
    self._ThreadProc()
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/__init__.py", line 160, in _ThreadProc
    self._ExecuteTestPhases(executor)
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/__init__.py", line 218, in _ExecuteTestPhases
    for phase_outcome in executor.ExecutePhases():
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_executor.py", line 170, in ExecutePhases
    result = self._ExecuteOnePhase(self.test_state.pending_phases[0])
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_executor.py", line 194, in _ExecuteOnePhase
    outcome_wrapper.SetOutcome(phase_thread.JoinOrDie())
  File "build/bdist.linux-x86_64/egg/contextlib2.py", line 97, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/phase_data.py", line 143, in RecordPhaseTiming
    values = dict(test_state.phase_data.measurements)
TypeError: 'NoneType' object is not iterable

Traceback (most recent call last):
  File "examples/foo.py", line 93, in <module>
    test.Execute(test_start=triggers.PromptForTestStart())
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/__init__.py", line 188, in Execute
    record = self._executor.GetState().GetFinishedRecord()
  File "/home/wallaceb/source/venv/local/lib/python2.7/site-packages/openhtf-0.9-py2.7.egg/openhtf/exe/test_state.py", line 160, in GetFinishedRecord
    'Internal framework error, test outcome unset!')
openhtf.exe.test_state.FrameworkError: Internal framework error, test outcome unset!

**Comment by jettisonjoe** [comment]
Ah, I see. If it's being run with `_ExecuteOnePhase` it's not going to work in the exit handler. The easiest way is probably to just leave it where you had it if that seems to be working.

File: openhtf/exe/__init__.py (hunk 2 above)
Comment (wallacbe) [inline openhtf/exe/__init__.py]: removed pending phases, added check for empty list around the pop in phase executor

**Comment by wallacbe** [comment]
PTAL

**Comment by jettisonjoe** [comment]
LGTM, assuming tests pass. Need second approval (from @grybmadsci) to merge.

File: openhtf/exe/__init__.py [hunk 3]
```diff
@@ -66,9 +67,14 @@ class TestExecutor(threads.KillableThread):
                          ['CREATED', 'START_WAIT', 'INITIALIZING', 'EXECUTING',
                           'STOP_WAIT', 'FINISHING'])
 
-  def __init__(self, test_data, plug_manager):
+  def __init__(self, test_data, plug_manager, teardown_function=None):
     super(TestExecutor, self).__init__(name='TestExecutorThread')
 
+    if teardown_function:
```
Comment (grybmadsci) [inline openhtf/exe/__init__.py]: `self._teardown_function = teardown_function or openhtf.PhaseInfo.WrapOrCopy(teardown_function)`

File: examples/all_the_things.py [hunk 4]
```diff
@@ -109,5 +111,5 @@ def attachments(test):
   #  with open('my_private_key.json', 'r') as json_file:
   #    test.AddOutputCallbacks(mfg_inspector.UploadToMfgInspector.from_json(
   #        json.load(json_file)))
-
+  test.Configure(teardown_function=teardown)
```
Comment (grybmadsci) [inline examples/all_the_things.py]: if you're going to add a call to `.Configure()` here, might as well also disable `http_api` with `http_port=None`.

**Comment by grybmadsci** [comment]
Nice, cleanly implemented, I like it :)  couple super minor nits, LGTM after that.

File: examples/all_the_things.py (hunk 4 above)
Comment (wallacbe) [inline examples/all_the_things.py]: Done

File: openhtf/exe/__init__.py (hunk 3 above)
Comment (wallacbe) [inline openhtf/exe/__init__.py]: Done

**Comment by wallacbe** [comment]
Made changes PTAL

**Pull request merged**